Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrMetadataEmitter.kt

```diff
@@ -0,0 +1,333 @@
+        }.toUniqId()
+    }
+
+    fun uniqIdForProperty(property: PropertyStub): UniqId = with (mangler) {
+        when (property.origin) {
+            is StubOrigin.ObjCProperty -> {
+                require(this.context is ManglingContext.Entity) {
+                    "Unexpected mangling context $context for property ${property.name}."
+                }
+                property.origin.property.uniqueSymbolName
+            }
+            is StubOrigin.Constant -> property.origin.constantDef.uniqueSymbolName
+            is StubOrigin.Global -> property.origin.global.uniqueSymbolName
+            // TODO: What to do with origin for enum entries and struct fields?
+            else -> error("Unexpected origin ${property.origin} for property ${property.name}.")
+        }.toUniqId()
+    }
+
+    fun uniqIdForTypeAlias(typeAlias: TypealiasStub): UniqId = with (mangler) {
+        when (typeAlias.origin) {
+            is StubOrigin.TypeDef -> typeAlias.origin.typedefDef.uniqueSymbolName
+            is StubOrigin.Enum -> typeAlias.origin.enum.uniqueSymbolName
+            else -> error("Unexpected origin ${typeAlias.origin} for typealias ${typeAlias.alias.fqName}.")
+        }
+    }.toUniqId()
+
+    private fun String.toUniqId() = UniqId(cityHash64())
+}
+
+/**
+ * Translates single [StubContainer] to [KmModuleFragment].
+ */
+internal class ModuleMetadataEmitter(
+        private val packageFqName: String
+) : StubIrToMetadataTranslationExtensions, StubIrVisitor<StubContainer?, Unit> {
+
+    private val uniqIds = StubIrUniqIdProvider()
+
+    private val classes = mutableListOf<KmClass>()
+    private val properties = mutableListOf<KmProperty>()
+    private val typeAliases = mutableListOf<KmTypeAlias>()
+    private val functions = mutableListOf<KmFunction>()
+
+    fun writeModule(): KmModuleFragment = KmModuleFragment().also { km ->
+        km.fqName = packageFqName
+        km.classes.addAll(classes)
+        km.pkg = writePackage()
+    }
+
+    private fun writePackage() = KmPackage().also { km ->
+        km.fqName = packageFqName
+        km.typeAliases.addAll(typeAliases)
+        km.properties.addAll(properties)
+        km.functions.addAll(functions)
+    }
+
+    override fun visitClass(element: ClassStub, data: StubContainer?) {
+        // TODO("not implemented")
+    }
+
+    override fun visitTypealias(element: TypealiasStub, data: StubContainer?) {
+        KmTypeAlias(element.flags, element.alias.topLevelName).apply {
+            uniqId = uniqIds.uniqIdForTypeAlias(element)
+            underlyingType = element.aliasee.map()
+            expandedType = element.aliasee.expandedType.map()
+        }.let(typeAliases::add)
+    }
+
+    override fun visitFunction(element: FunctionStub, data: StubContainer?) {
+        KmFunction(element.flags, element.name).apply {
+            annotations.addAll(element.annotations.map { it.map() })
```

Comment:
If `annotations` is a collection, then consider using `element.annotations.mapTo(annotations) { ... }` instead.
Same for other similar code fragments.